File: nodes/core/market/filters/ema_range_filter_node.py

import pandas as pd
import numpy as np
from typing import List
from ta.trend import EMAIndicator
from nodes.core.market.filters.base.base_indicator_filter_node import BaseIndicatorFilter
from core.types_registry import IndicatorResult, IndicatorType, OHLCVBar, IndicatorValue

class EmaRangeFilter(BaseIndicatorFilter):
    """
    Filters assets where EMA(10, high-low range) > close / 100
    """

    default_params = {
        "timeperiod": 10,
        "divisor": 100.0,
    }

    params_meta = [
        {"name": "timeperiod", "type": "number", "default": 10, "min": 1, "step": 1},
        {"name": "divisor", "type": "number", "default": 100.0, "min": 1.0, "step": 1.0},
    ]

    def _validate_indicator_params(self):
        if self.params["timeperiod"] < 1:
            raise ValueError("Time period must be at least 1")
        if self.params["divisor"] <= 0:
            raise ValueError("Divisor must be positive")

    def _calculate_indicator(self, ohlcv_data: List[OHLCVBar]) -> IndicatorResult:
        if not ohlcv_data:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA_RANGE,
                values=IndicatorValue(lines={"ema_range": np.nan, "close": np.nan}),
                params=self.params,
                error="No data"
            )

        df = pd.DataFrame(ohlcv_data)
        if len(df) < self.params["timeperiod"]:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA_RANGE,
                timestamp=int(df['timestamp'].iloc[-1]) if 'timestamp' in df else 0,
                values=IndicatorValue(lines={"ema_range": np.nan, "close": df['close'].iloc[-1]}),
                params=self.params,
                error="Insufficient data"
            )

        price_range = df['high'] - df['low']
        ema_indicator = EMAIndicator(price_range, window=self.params["timeperiod"])
        ema_series = ema_indicator.ema_indicator()
        latest_ema = ema_series.iloc[-1] if not ema_series.empty else np.nan
        latest_close = df['close'].iloc[-1]

        values = IndicatorValue(lines={
            "ema_range": latest_ema if not pd.isna(latest_ema) else np.nan,
            "close": latest_close
        })

        return IndicatorResult(
            indicator_type=IndicatorType.EMA_RANGE,
            timestamp=int(df['timestamp'].iloc[-1]) if 'timestamp' in df else 0,
            values=values,
            params=self.params
        )

    def _should_pass_filter(self, indicator_result: IndicatorResult) -> bool:
        if indicator_result.error or not hasattr(indicator_result.values, 'lines'):
            return False

        values = indicator_result.values.lines
        if "ema_range" not in values or "close" not in values:
            return False

        ema_range = values["ema_range"]
        close = values["close"]
        if pd.isna(ema_range) or pd.isna(close):
            return False

        return ema_range > close / self.params["divisor"]

File: nodes/market/filters/ema-range-filter-node.ts
// src/nodes/core/market/filters/ema-range-filter-node.ts
// Translated from: nodes/core/market/filters/ema_range_filter_node.py

import { BaseIndicatorFilter } from './base/base-indicator-filter-node';
import { IndicatorType, createIndicatorResult, createIndicatorValue } from '@fig-node/core';
import type {
  ParamMeta,
  DefaultParams,
  OHLCVBar,
  IndicatorResult,
  NodeUIConfig,
} from '@fig-node/core';
import { calculateEma } from '../calculators/ema-calculator';

/**
 * Filters assets where EMA(10, high-low range) > close / 100
 */
export class EmaRangeFilter extends BaseIndicatorFilter {
  static override defaultParams: DefaultParams = {
    timeperiod: 10,
    divisor: 100.0,
  };

  static override paramsMeta: ParamMeta[] = [
    { name: 'timeperiod', type: 'number', default: 10, min: 1, step: 1 },
    { name: 'divisor', type: 'number', default: 100.0, min: 1.0, step: 1.0 },
  ];

  static uiConfig: NodeUIConfig = {
    size: [220, 100],
    displayResults: false,
    resizable: false,
  };

  protected override validateIndicatorParams(): void {
    const timeperiod = this.params.timeperiod ?? 10;
    const divisor = this.params.divisor ?? 100.0;

    if (typeof timeperiod !== 'number' || timeperiod < 1) {
      throw new Error('Time period must be at least 1');
    }
    if (typeof divisor !== 'number' || divisor <= 0) {
      throw new Error('Divisor must be positive');
    }
  }

  protected calculateIndicator(ohlcvData: OHLCVBar[]): IndicatorResult {
    if (!ohlcvData || ohlcvData.length === 0) {
      return createIndicatorResult({
        indicatorType: IndicatorType.EMA_RANGE,
        timestamp: 0,
        values: createIndicatorValue({ single: NaN, lines: { ema_range: NaN, close: NaN } }),
        params: this.params,
        error: 'No data',
      });
    }

    const timeperiodParam = this.params.timeperiod;
    const timeperiod = Math.floor(typeof timeperiodParam === 'number' ? timeperiodParam : 10);

    if (ohlcvData.length < timeperiod) {
      return createIndicatorResult({
        indicatorType: IndicatorType.EMA_RANGE,
        timestamp: ohlcvData[ohlcvData.length - 1].timestamp,
        values: createIndicatorValue({ single: NaN, lines: { ema_range: NaN, close: ohlcvData[ohlcvData.length - 1].close } }),
        params: this.params,
        error: 'Insufficient data',
      });
    }

    // Calculate price range
    const priceRange = ohlcvData.map((bar) => bar.high - bar.low);
    const closes = ohlcvData.map((bar) => bar.close);

    // Calculate EMA using the calculator
    const emaResult = calculateEma(priceRange, timeperiod);
    const emaValues = emaResult.ema;
    const latestEma = emaValues.length > 0 ? emaValues[emaValues.length - 1] : null;
    const latestClose = closes[closes.length - 1];

    const emaRangeValue = latestEma !== null ? latestEma : NaN;

    return createIndicatorResult({
      indicatorType: IndicatorType.EMA_RANGE,
      timestamp: ohlcvData[ohlcvData.length - 1].timestamp,
      values: createIndicatorValue({
        single: NaN,
        lines: {
          ema_range: emaRangeValue,
          close: latestClose,
        },
      }),
      params: this.params,
    });
  }

  protected shouldPassFilter(indicatorResult: IndicatorResult): boolean {
    if (indicatorResult.error || !indicatorResult.values.lines) {
      return false;
    }

    const values = indicatorResult.values.lines;
    if (!('ema_range' in values) || !('close' in values)) {
      return false;
    }

    const emaRange = values.ema_range;
    const close = values.close;
    if (Number.isNaN(emaRange) || Number.isNaN(close)) {
      return false;
    }

    const divisorParam = this.params.divisor;
    const divisor = typeof divisorParam === 'number' ? divisorParam : 100.0;

    return emaRange > close / divisor;
  }
}
